runtime/compiler/env/J9SharedCache.cpp

```diff
@@ -813,76 +829,155 @@ TR_J9SharedCache::findChainForClass(J9Class *clazz, const char *key, uint32_t ke
+      }
+   else if (chainPtr != chainEnd)
+      {
+      LOG(1, "\tfinished classes and interfaces, but not at chain end, returning false\n");
+      }
+   else
+      {
+      validationSucceeded = true;
+      }
+
+   return validationSucceeded;
+   }
+
 bool
 TR_J9SharedCache::classMatchesCachedVersion(J9Class *clazz, UDATA *chainData)
    {
    J9ROMClass *romClass = TR::Compiler->cls.romClassOf(fe()->convertClassPtrToClassOffset(clazz));
    J9UTF8 * className = J9ROMCLASS_CLASSNAME(romClass);
-   LOG(5, { log("classMatchesCachedVersion class %p %.*s\n", clazz, J9UTF8_LENGTH(className), J9UTF8_DATA(className)); });
+   LOG(1, "classMatchesCachedVersion class %p %.*s\n", clazz, J9UTF8_LENGTH(className), J9UTF8_DATA(className));
 
    uintptr_t classOffsetInCache;
+
+   /* If the pointer isn't the SCC, then return false immmediately
+    * as the bit vectors hold offsets into the SCC of romclasses
+    */
    if (!isPointerInSharedCache(romClass, &classOffsetInCache))
       {
-      LOG(5, { log("\tclass not in shared cache, returning false\n"); });
+      LOG(1, "\tclass not in shared cache, returning false\n");
       return false;
       }
 
+   /* Check if the validation of the class chain was previously
+    * performed; if so, return the result of that validation
+    */
+   if (TR::Options::getAOTCmdLineOptions()->getOption(TR_EnableClassChainValidationCaching))
+      {
+      OMR::CriticalSection checkBitVectors(classChainValidationMutex);
+      if (_classChainValidationFailed->isSet(classOffsetInCache))
+         {
+         LOG(1, "\tcached result: validation failed\n");
+         return false;
+         }
+      if (_classChainValidationSucceeded->isSet(classOffsetInCache))
+         {
+         LOG(1, "\tcached result: validation succeeded\n");
+         return true;
+         }
+      }
+
+   /* If the chainData passed in is NULL, try to find it in the SCC
+    * using the romclass
+    */
    if (chainData == NULL)
       {
       char key[17]; // longest possible key length is way less than 16 digits
       uint32_t keyLength;
       createClassKey(classOffsetInCache, key, keyLength);
-      LOG(9, { log("\tno chain specific, so looking up for key %.*s\n", keyLength, key); });
+      LOG(3, "\tno chain specific, so looking up for key %.*s\n", keyLength, key);
       chainData = findChainForClass(clazz, key, keyLength);
-      if (chainData == NULL)
+      }
+
+   /* If the chainData is still NULL, add the romclas offset to
+    * the _classChainValidationFailed bit vector
+    */
+   if (chainData == NULL)
+      {
+      LOG(1, "\tno stored chain, returning false\n");
+      if (TR::Options::getAOTCmdLineOptions()->getOption(TR_EnableClassChainValidationCaching))
          {
-         LOG(5, { log("\tno stored chain, returning false\n"); });
-         return false;
+         OMR::CriticalSection noStoredChain(classChainValidationMutex);
+         _classChainValidationFailed->set(classOffsetInCache);
```

Comment:
Each class offset is represented by a bit inside the bit-vector. You have 1 million bits (initial size) which means you can cover (more or less) an SCC size of 1MB (ignoring for a moment that classes are stored in the beginning of the SCC).
If a class is always aligned on an 8-byte boundary you could do (classOffsetInCache >> 3) to expand your coverage

Reply:
> You have 1 million bits (initial size) which means you can cover (more or less) an SCC size of 1MB 

I believe it's going to be much more than 1MB. The reason is because the ROMClass offsets refer to the embedded ROMClass in the SCC. The ROMClass contains the ROMConstantPool and the ROMMethods (which contain all the bytecodes). Therefore, the distance between two ROMClasses is going to much more than 1 pointer size. 

> If a class is always aligned on an 8-byte boundary you could do (classOffsetInCache >> 3) to expand your coverage

I don't think the ROMClasses are not aligned on a 8-byte boundary like RAMClasses, but i'll defer to @DanHeidinga on that. However even if they were aligned, I don't believe you can do the shift because the shift only applies to the class and not the offset.

Reply:
The distance between two ROMClasses is indeed going to be larger than 1 pointer size which is why the bitvector solution is inefficient. A sparse bit vector may be better here.

If the ROMClasses are aligned on a 8-byte boundary, the offsets are going to be aligned too because I assume that the section of ROMclasses in the SCC is aligned.

Reply:
Resolving thread based on the conversation starting from https://github.com/eclipse/openj9/pull/9403#issuecomment-623618092